kep/real-load-aware-scheduling/real-load-aware-scheduling.md

```diff
@@ -0,0 +1,540 @@
+4. There is a break in the graph with a high drop due to the "penalty" we have.
+
+#### Spreading with BestFitBinPack
+
+
+To meet the 2nd constraint in the problem statement, we utilize the existing PodTopologySpread (PTS) plugin provided by K8s. It was proved with experiments that using a dynamic weight-based approach is not a good idea here. However, if we choose a fixed large weight for PTS, say 100, we can meet our requirements. The problem with this approach is that we will end up downplaying the scores of other important default plugins like Image Locality, Interpod Affinity, etc. with their default weights. Also, with more plugins we plan to add in the future, managing weights will not be simple. So for BFBP to work well with PTS without modifying weights, the following extended algorithm is proposed that is called when scheduling replicas of a service, replica-set, replication controller, etc.
+
+
+**Algorithm**
+
+
+
+1. For each node, calculate S = (1000*PTS + PTS*BFBP)
+2. Scale each S to be between 0-100
+3. Return the scores
+
+**Algorithm Analysis**
+
+1. The algorithm favors nodes with high PTS scores with 1000*PTS part. The weight 1000 is chosen to give preference to PTS  and to avoid the congestion of scores upon scaling down, my magnifying the scores 
+2. PTS*BFBP penalizes hot nodes with low BFBP scores.
+3. The final scaled-down value deals with congestion, hot node penalty, and favoring of high PTS score nodes
+
+Example: 
+
+
+| Node | PTS | BFBP | Combined Score |
+|------|-----|------|----------------|
+|  N1  | 100 |  35  |       96       |
+|  N2  |  67 |  75  |       67       |
+|  N3  | 100 |  80  |       100      |
+|  N4  |  82 |  90  |       83       |
+|  N5  |  0  |  90  |        0       |
+
+
+
+### Safe Balancing Plugin
+
+
+#### PreScore plugin
+
+This extension point, which is invoked for all the nodes at once, will be used to cache metrics for all nodes using a single REST API  call. This optimization will reduce multiple calls for each node from the Score plugin below, reducing overall scheduling latency.
+
+
+#### Score plugin
+
+This plugin would extend the Score extension point. K8s scheduler framework calls the Score function for each node separately when scheduling a pod.
+
+
+**Idea**
+
+
+Balancing load based on average would be risky sometimes, as it does not consider the bursty variations. A safe balancing plugin balances not only the average load but also the risk caused by load variations. Suppose we take the mean (M) and standard deviation (V) of all nodes’ utilization into a mu-sigma plot below. In that case, the safe balancing plugin will make placements such that all nodes’ utilization are aligned on the diagonal line, which is V + M = c. Here, c is a constant indicating the overall cluster utilization average plus the standard deviation.
```

Comment:
Where do `M` and `V` come from? i.e., directly from LoadWatcher? or based on the raw data of LoadWather plus extra calculation? It'd good to use a concrete flow to explain how these values are generated, and hence make readers understand their semantics easily.

Reply:
Directly from load watcher. I will explain this under watcher and add an example json payload in appendix.